Implement cl_khr_command_buffer_multi_device

jansol opened the pull request:
With command buffers being implemented "host-side", this is a fairly straightforward thing to support.

jansol commented:
The formatting scripts had gotten out of sync with each other recently so this also brings those back in line.

pjaaskel commented on lib/CL/clRemapCommandBufferKHR.c:
```diff
@@ -59,23 +59,29 @@ POname (clRemapCommandBufferKHR) (
   POCL_GOTO_ERROR_COND((universal_sync == 0 && num_queues > 1), CL_INCOMPATIBLE_COMMAND_QUEUE_KHR);
 
-  POname (clCreateCommandBufferKHR) (num_queues, queues, command_buffer->properties, &errcode);
+  new_cmdbuf = POname (clCreateCommandBufferKHR) (num_queues, queues, command_buffer->properties, &errcode);
```
Format? Long line.

pjaaskel commented on tests/runtime/test_command_buffer_multi_device.c:
```diff
@@ -303,7 +303,7 @@ main (int _argc, char **_argv)
     }
 
-  /* Remap from 2 to 1 queues & run */
+  /* Remap from N to 1 queues & run */
```
Can we test the opposite case as well? That is, the one that can enable runtime-driven task mapping/perf. portability via pushing all work to a default CPU queue and then remapping to utilize more devices. That likely is a NOP for now. Also it would be good to test a case where the remap is done to queue/devices which are not in the original schedule/buffer. This models the disappearing devices case.

pjaaskel commented on lib/CL/clCreateCommandBufferKHR.c:
```diff
@@ -96,6 +104,8 @@ POname (clCreateCommandBufferKHR) (
         }
     }
 
+  POCL_GOTO_ERROR_COND ((num_queues > 1 && !pocl_cmdbuf_can_queues_sync(num_queues, queues, universal_sync_enabled)), CL_INCOMPATIBLE_COMMAND_QUEUE_KHR);
```
Long lines here as well.

pjaaskel commented on lib/CL/pocl_util.c:
```diff
@@ -1240,6 +1240,43 @@ pocl_cmdbuf_get_property (cl_command_buffer_khr command_buffer,
+          cl_device_id dev = queues[i]->device;
+
+          cl_uint num_sync_devices;
+          POCL_RETURN_ERROR_COND ((POname(clGetDeviceInfo)(queues[i]->device, CL_DEVICE_COMMAND_BUFFER_NUM_SYNC_DEVICES_KHR, sizeof (cl_uint), &num_sync_devices, NULL)), CL_INCOMPATIBLE_COMMAND_QUEUE_KHR);
```
Here too.

pjaaskel reviewed (changes requested):
Generally looks good, but can you run the formatter (at least long lines pop to eye) and add a bit of comments to the overall functioning of the current remap()? Also perhaps squash some of the commits together if it's not too much trouble (I accidentally commented some obsolete issues related to multiple devices, I think, when going through the commits).

pjaaskel commented:
Can you describe this in the release notes and rebase?

jansol commented:
Rebased, squashed & formatted. Should be easier to review now. I'm seeing a number of asan warnings though (locally as well), still working through those.

pjaaskel commented on doc/sphinx/source/notes_6_0.rst:
```diff
@@ -20,6 +20,15 @@ any particular behavior (or lack thereof) they are treated as a no-op. However
+Multi-device command buffers
+============================
+
+Initial support for `cl_khr_command_buffer_multi_device` has been added, i.e. it
```
For the CPU devices, right?

jansol replied:
There is not really anything driver-specific, just like in the original command buffer code. It just hasn't really been tested with other drivers.

pjaaskel replied:
Perhaps still worth mentioning what are the expectations in terms of perf. improvements (none still, right?) as it "caches" mainly the error checking?

pjaaskel commented:
The arm bots are broken again, pulling this in regardless.
